### ISSUE TITLE
Ant Design currently generates inaccessible forms. Blind people cannot use forms made by Ant Design.

### ISSUE THREAD

**Issue by thisissami** [ant-design/ant-design#16270]
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
A substantial percentage of the internet relies on screen readers (e.g. all blind people, many old people, etc.) - and having forms be easily read by them is critical. People who use screen readers currently *cannot* use forms generated by Ant Design. 

Considering how widely adopted And Design is, many people are trying to access Ant Design forms with screen readers daily - and currently, they can't read them. AntD has been advertised as an accessible platform, and accordingly - I hope that it is updated to make forms accessible. 

Here is a post with a lot more info (including screenshots of current problems): https://github.com/ant-design/ant-design/issues/16268

### What does the proposed API look like?
The proposed API would be the same as it is currently. However, the output would be very slightly different. All that's needed is 2 extra attributes per `<Form.Item/>` - e.g. `for` in label, and `id` in input. That's really all that's needed for all the Ant Design forms to become accessible!!

As an alternative, perhaps the top level `<Form />` can have an extra prop available called `accessible` that is a boolean, and when set to `true`, forms are made accessible. (personally, however, i think they should be accessible by default always).

-----------------------------------------
* [ ] Select
* [ ] Checkboxes
* [ ] Radio Buttons
* [ ] Datepicker
* [ ] Modal
* [ ] Menu

description ref: [comment](#issuecomment-486393323)

**Comment by thisissami** [comment]
It seems that when using `const { getFieldDecorator } = form;` and 

```javascript
<Form.Item label="Title">
              {getFieldDecorator('title', {
                rules: [{ required: true, message: 'Please input the title of collection!' }],
              })(
                <Input />
              )}
            </Form.Item>
```

within the `<Form/>` - proper accessibility kicks in (unless you wrap an `<Input />` with an `<Autocomplete />`). However, one shouldn't be forced to use the state management in order to have good accessible forms.

**Comment by zombieJ** [comment]
Do you mean this? https://codesandbox.io/s/48m15l1q5w

**Comment by thisissami** [comment]
Yes, like that. I think you guys should make it clear that this is possible in the documentation for `Form.Item`. it would make a big difference for people who care about this. :) 

However, there's still some issues:

I played around with this some more. Here's the same thing with 5 different types of inputs: https://codesandbox.io/s/5x02k0443p With the `DatePicker`, the `id` that the `label` points to is on the incorrect element (`Tota11y` still complains about not having a label available for that input). With the `CheckboxGroup`, antd complains about the `id` prop not existing. 

I haven't done enough tests with this (I'll need to actually run the outputted HTML from this code using a Screen Reader to really understand how well it works for blind people), but my immediate sense is that the basic input is good - but other inputs for forms could still be problematic. 

I'd be happy to do some tests tomorrow & report back! :) Thanks for the first bit of info on this @zombieJ

**Comment by zombieJ** [comment]
Thanks for your help. I will add it in document. Maybe you can help to add list in the issue above : )

**Comment by thisissami** [comment]
@zombieJ what do you mean by `help to add list in the issue above`? 

An update, in using VoiceOver with Opera:

The input with text is good, and easy to understand. Unfortunately, it (and buttons) are the only part of AntD forms that are truly accessible to understand. 

I tested my form set up in a website that uses antd `Menu`, with the form embedded in an antd `Modal`, and discovered accessibility issues with both of those components as well. Here are some of the problems that I've encountered so far:

*Form*

1) Select - the screen reader only reads the first item from the options. It gives zero sense that this is a select, and what type of selection it is. 

2) Checkboxes & Radio Buttons - these share the same problem. An individual radio button or checkbox is accessible. However, there is no connection to the label. The screenreader can't understand what the relationship is between all the different checkboxes/radio buttons. 

For example, if there was a "Do you like cake?" with 2 radio options of "yes" and "no", the screen reader would be able to tell you that you had a yes and no, and which one was selected - but you wouldn't know what the yes & no were about. 

3) Datepicker - it is clear that this is something to pick a date, and the screenreader tells you that you can enter text. However, it is impossible to interact with the actual date picker (only entering text is easy). I can imagine that picking from a calendar with a screenreader is really really hard to design, so I don't think this can be a large concern. 

*Modal*

4) For some reason, antd is adding a `sentinalStart` and `sentinalEnd` before/after all the other elements of the Modal. This is read by the screenreader, and is the first thing that is read to you - and makes the user experience very very confusing. It's hidden visibly by html, but the text is still accessible. Here's an example:

```html
<div tabindex="0" style="width: 0px; height: 0px; overflow: hidden;">sentinelStart</div>
```
It's the first thing within the `ant-modal` `div`. 

5) It is impossible to navigate a form within a modal using the screenreader. The screenreader can tell me what's on the screen, but it can't actually navigate within the layout. There's something weird about how it's structured. That being said - it *is* possible to navigate using the tab button, and that's very important. (But the tab has 2 extra stops - the `sentinalStart` and `sentinalEnd`)

6) A modal with a form should have an autofocus to the first field in the form, and when it closes, should autofocus back to the button/link that opened the modal. Is this straightforward to do with Ant design's current HTML? 

*Menu*

7) In using a screenreader to test this, I discovered that `Menu.Item` is being read as a form element by the Screen Reader, when it really shouldn't be. It makes it much harder to navigate a website, as the screen reader tells you there are form elements when there aren't. It lists each of them as "Dashboard menu item", "account menu item", etc. so it is clear that it's a menu item, but it still makes the user experience with a screen reader very difficult, and one that never happens on other websites. 

I hope this is all useful, and I hope that the Ant Design team addresses this. :) You've made such a wonderful framework, and it would be great to put in little changes to allow several million more people to be able to use websites created with Ant Design! :)

**Comment by zombieJ** [comment]
Thanks for your help! We will continually working on this. Let me update it as check list.

**Comment by thisissami** [comment]
You're welcome @zombieJ - i'm happy to help! I'll update here with more if I find more. :)

**Comment by Stephen2** [comment]
❤️ love the commitment to improving accessibility.  Antd has already merged 3+ accessibility improvements I've suggested, over just a few weeks.

Thank you for continuing to be so awesome

**Comment by thisissami** [comment]
Me too! :) I really love using Antd, and unfortunately my team would have to switch to a different framework if accessibility wasn't a priority. I'm hopeful that the suggestions in this thread get implemented in the next month or so, otherwise we're gonna have to seriously start thinking about alternative frameworks. :(

**Comment by zombieJ** [comment]
@thisissami , since some components do not works well with accessibility. We should adjust some dom structure to enhance it. Considering not breaks current behavior, it may not fast for it but it's in progress.

**Comment by thisissami** [comment]
Thanks for the update @zombieJ. I appreciate that the antd team is working on this.

**Comment by thisissami** [comment]
Here is a react framework that primarily focuses on accessibility: https://reakit.io/ / https://github.com/reakit/reakit it may be a good reference as `antd` tries to improve it's accessibility.

**Comment by andreobriennz** [comment]
I was about to report this as an issue and then I came across this

I've also noticed that the expandable rows in the Table component aren't accessible. There is no way to expand them via the keyboard

Here's an example:
https://codesandbox.io/embed/trusting-kowalevski-7zmm4

**Comment by cdeutsch** [comment]
> Select - the screen reader only reads the first item from the options. It gives zero sense that this is a select, and what type of selection it is.

I was looking into this and I notice that [react-responsive-select](https://github.com/benbowes/react-responsive-select) gives their `Option`'s the role `menuitem` vs Ant which uses the role `option`, and the container is `menu` vs `listbox`

If you use Voice Over on Mac on just a [plain old `<select>`](https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_select) it also seems to treat the `<option>`'s as `menuitem` which is counter intuitive.

I did try updating Ant to use `menu` and `menuitem` but it wasn't enough to get Voice Over to read out the options as they're selected. So I'm not sure what `react-responsive-select` is doing to make it work.

**Comment by kaushalyap** [comment]
Any progress with this? Accessibility matters

**Comment by afc163** [comment]
@kaushalyap PR is welcome~

**Comment by Gibbo3771** [comment]
Issues with this here as well. Can't add Aria Labels to elements?

**Comment by dgreene1** [comment]
@andreobriennz, [your comment](https://github.com/ant-design/ant-design/issues/16270#issuecomment-509479012) seems to be another issue. Can you open a new issue with AntD? I'd like to follow that one as well.

**Comment by afc163** [comment]
- [ ] Table: https://github.com/ant-design/ant-design/issues/11194 #18476

**Comment by Skyerin** [comment]
Just to add to this conversation about inaccessible forms, we also don't have roles set on the Radio.Group component. Without this, screenreaders cannot properly recognise the radio group. The role on Radio.Group should be 'radiogroup'.

**Comment by aalvarezwindey** [comment]
> Do you mean this? https://codesandbox.io/s/48m15l1q5w

Hello! @zombieJ 

Is it possible to generate an accessible input but hide the label?

For example: 
![image](https://user-images.githubusercontent.com/45921171/225298714-6eb48a1a-96bd-4cc1-b52b-9edc531c8986.png)

First name input is accessible and Last name input not... In our designs we are not using labels, just placeholders.

I tried `hidden=true` in `Form.Item` but it hides the input itself also

**Comment by aalvarezwindey** [comment]
> > Do you mean this? https://codesandbox.io/s/48m15l1q5w
> 
> Hello! @zombieJ
> 
> Is it possible to generate an accessible input but hide the label?
> 
> For example: ![image](https://user-images.githubusercontent.com/45921171/225298714-6eb48a1a-96bd-4cc1-b52b-9edc531c8986.png)
> 
> First name input is accessible and Last name input not... In our designs we are not using labels, just placeholders.
> 
> I tried `hidden=true` in `Form.Item` but it hides the input itself also

A workaround with styled-components class override for whom it is useful :smile: 

```jsx
import { Form } from 'antd';
import styled, { css } from 'styled-components';

type AntdFormItemProps = React.ComponentProps<typeof Form.Item>;
type CustomFormItemProps = {
  hideLabel?: boolean;
};

type FormItemProps = AntdFormItemProps & CustomFormItemProps;

// Adding label makes items more accessible
const hideLabelCss = css`
  .ant-form-item-label {
    display: none;
  }
`;

export const FormItem: React.FC<FormItemProps> = styled(Form.Item)<FormItemProps>`
  ${({ hideLabel = true }) => (hideLabel ? hideLabelCss : '')}
`;

```

**Comment by JulesLecuir** [comment]
Hi, any news on that ? Thank you !

**Comment by vorant94** [comment]
just one additional thing to mention is that accessibility isn't only about blind people and others like that. it is also about easier automation development for e2e tests with tools like playwright... I was surprised that color picker of library with 1.5kk downloads is just a bunch of div's

although I have to say there isn't much libraries with such huge amount of components as antd especially free ones, so anyway thanks for the work you are doing there!
